File: backend/app/schemas/lv.py
from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


class LVCreate(BaseModel):
    name: str
    trade: str
    vorbemerkungen: str | None = None


class LVUpdate(BaseModel):
    name: str | None = None
    status: str | None = None
    vorbemerkungen: str | None = None


class PositionUpdate(BaseModel):
    """Partial update payload for a single position.

    All fields optional → caller PATCH-style sends only what changed.
    ``menge`` is exposed for the v23.5 inline-edit flow so reviewers
    can override calculated quantities without re-running the engine
    (e.g. when the AI math is right but the room input was wrong and
    re-running would be slower than typing the corrected value).
    Locked positions reject every field — that gate lives at the
    endpoint layer so all writes go through one check, not five.
    """

    kurztext: str | None = None
    langtext: str | None = None
    menge: float | None = None
    einheitspreis: float | None = None
    is_locked: bool | None = None


class BerechnungsnachweisResponse(BaseModel):
    """Traceable calculation proof for a single position × room pair.

    The underlying DB columns retain their legacy ``onorm_*`` names
    because renaming them would trigger an unnecessary migration and
    the names are nowhere user-visible. We expose them under generic
    public names (``rule_factor``, ``rule_ref``, ``rule_paragraph``)
    so the API surface stays norm-neutral — the application is a
    calculation engine, not a standards reference.

    ``populate_by_name`` + ``Field(alias=...)`` means the ORM row's
    ``onorm_factor`` attribute still fills the ``rule_factor`` field
    without a manual projection layer; the response JSON uses the
    generic keys exclusively.
    """

    model_config = ConfigDict(from_attributes=True, populate_by_name=True)

    id: UUID
    position_id: UUID
    room_id: UUID
    raw_quantity: float
    formula_description: str
    formula_expression: str
    rule_factor: float = Field(alias="onorm_factor")
    rule_ref: str | None = Field(default=None, alias="onorm_rule_ref")
    rule_paragraph: str | None = Field(default=None, alias="onorm_paragraph")
    deductions: list[dict]
    net_quantity: float
    unit: str
    notes: str | None


class PositionResponse(BaseModel):
    id: UUID
    gruppe_id: UUID
    positions_nummer: str
    kurztext: str
    langtext: str | None
    einheit: str
    menge: float | None
    einheitspreis: float | None
    gesamtpreis: float | None
    positionsart: str
    text_source: str
    is_locked: bool
    sort_order: int
    berechnungsnachweise: list[BerechnungsnachweisResponse] = []

    model_config = {"from_attributes": True}


class LeistungsgruppeResponse(BaseModel):
    id: UUID
    lv_id: UUID
    nummer: str
    bezeichnung: str
    sort_order: int
    positionen: list[PositionResponse] = []

    model_config = {"from_attributes": True}


class LVResponse(BaseModel):
    id: UUID
    project_id: UUID
    name: str
    trade: str
    status: str
    vorbemerkungen: str | None
    created_at: datetime
    updated_at: datetime
    gruppen: list[LeistungsgruppeResponse] = []

    model_config = {"from_attributes": True}

File: frontend/src/types/lv.ts
export interface Berechnungsnachweis {
  id: string;
  position_id: string;
  room_id: string;
  raw_quantity: number;
  formula_description: string;
  formula_expression: string;
  // Renamed from ``onorm_*`` in the backend response. The DB columns
  // keep the historic ``onorm_`` prefix for compatibility with existing
  // rows, but the API + UI use generic terms — the app is a calculation
  // engine, not a norm library.
  rule_factor: number;
  rule_ref: string | null;
  rule_paragraph: string | null;
  deductions: DeductionDetail[];
  net_quantity: number;
  unit: string;
  notes: string | null;
}

export interface DeductionDetail {
  opening: string;
  area: number;
  deducted: boolean;
  reason?: string;
}

export interface Position {
  id: string;
  gruppe_id: string;
  positions_nummer: string;
  kurztext: string;
  langtext: string | null;
  einheit: string;
  menge: number | null;
  einheitspreis: number | null;
  gesamtpreis: number | null;
  positionsart: string;
  text_source: string;
  is_locked: boolean;
  sort_order: number;
  berechnungsnachweise: Berechnungsnachweis[];
}

export interface Leistungsgruppe {
  id: string;
  lv_id: string;
  nummer: string;
  bezeichnung: string;
  sort_order: number;
  positionen: Position[];
}

export interface LV {
  id: string;
  project_id: string;
  name: string;
  trade: string;
  status: string;
  vorbemerkungen: string | null;
  created_at: string;
  updated_at: string;
  gruppen: Leistungsgruppe[];
}

export interface LVCreate {
  name: string;
  trade: string;
  vorbemerkungen?: string;
}
